src/mscorlib/shared/System/IO/Path.cs

```diff
@@ -80,16 +80,18 @@ public static string GetDirectoryName(string path)
                 throw new ArgumentException(SR.Arg_PathEmpty, nameof(path));
 
             path = PathInternal.NormalizeDirectorySeparators(path);
-            int root = PathInternal.GetRootLength(path);
+            int end = PathInternal.GetDirectoryNameOffset(path);
+            return end >= 0 ? path.Substring(0, end) : null;
+        }
 
-            int i = path.Length;
-            if (i > root)
-            {
-                while (i > root && !PathInternal.IsDirectorySeparator(path[--i])) ;
-                return path.Substring(0, i);
-            }
-            
-            return null;
+        public static ReadOnlySpan<char> GetDirectoryName(ReadOnlySpan<char> path)
+        {
+            if (PathInternal.IsEffectivelyEmpty(path))
+                return ReadOnlySpan<char>.Empty;
+
+            path = PathInternal.NormalizeDirectorySeparators(path);
```

Comment:
Ditto for above- the span APIs shouldn't allocate.

Reply:
should we change their string overloads to keep the normalization ? or should their behavior be same as the span overloads ?